File: lifeos/reviews/LIFEOS-P3-143/independent-re-review-4/review_tools/verify_final_manifest.py
#!/usr/bin/env python3
"""Verify the explicitly declared non-self-referential re-review-4 manifest."""

from __future__ import annotations

import hashlib
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent

def digest(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()

def main() -> None:
    manifest = json.loads((ROOT / "FINAL_MANIFEST.json").read_text(encoding="utf-8"))
    errors = []
    forbidden = {"FINAL_MANIFEST.json", "review_tools/verify_final_manifest.py", "manifest_verification.json"}
    for entry in manifest.get("entries", []):
        relative = entry.get("path", "")
        path = ROOT / relative
        if relative in forbidden or not path.is_file():
            errors.append(f"invalid_entry:{relative}")
        elif path.stat().st_size != entry.get("bytes"):
            errors.append(f"size_mismatch:{relative}")
        elif digest(path) != entry.get("sha256"):
            errors.append(f"hash_mismatch:{relative}")
    output = {"schema": "lifeos.p3-143.independent-rereview4.manifest-verification.v1", "entry_count": len(manifest.get("entries", [])), "errors": errors, "result": "PASS" if not errors else "FAIL"}
    (ROOT / "manifest_verification.json").write_text(json.dumps(output, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(json.dumps(output))
    raise SystemExit(0 if not errors else 1)

if __name__ == "__main__":
    main()

File: lifeos/reviews/LIFEOS-P3-143/independent-review/verify_final_manifest.mjs
import { readFile } from "node:fs/promises";
import { createHash } from "node:crypto";
import { resolve } from "node:path";

const root = import.meta.dirname;
const manifest = JSON.parse(await readFile(resolve(root, "FINAL_MANIFEST.json"), "utf8"));
const hash = (bytes) => createHash("sha256").update(bytes).digest("hex");
const errors = [];
if (manifest.schema !== "lifeos.p3-143.independent-review-final-manifest.v1" || manifest.task !== "LIFEOS-P3-143" || manifest.self_referential !== false) errors.push({ path: "FINAL_MANIFEST.json", error: "identity_or_self_reference" });
for (const forbidden of ["FINAL_MANIFEST.json", "verify_final_manifest.mjs", "manifest_verification.json"]) {
  if (manifest.files?.some((entry) => entry.path === forbidden)) errors.push({ path: forbidden, error: "self_inclusion" });
}
for (const entry of manifest.files ?? []) {
  try {
    const bytes = await readFile(resolve(root, entry.path));
    if (bytes.length !== entry.bytes || hash(bytes) !== entry.sha256) errors.push({ path: entry.path, error: "hash_or_size_mismatch" });
  } catch {
    errors.push({ path: entry.path, error: "missing" });
  }
}
const result = { schema: "lifeos.p3-143.independent-review-manifest-verification.v1", task: "LIFEOS-P3-143", entries: manifest.files?.length ?? 0, errors, result: errors.length ? "FAIL" : "PASS" };
console.log(JSON.stringify(result, null, 2));
if (errors.length) process.exitCode = 1;
